### PR TITLE
Food clusters

### PR DESCRIPTION
**Pull request by matthewr6**
find major clusters

### CODE REVIEW

**Comment by an-OK-squirrel** [comment]
I don't really feel like this would be good to merge. For me, it toggles from whereever it's going to a random point with no food, which makes progress slow. I don't know why this is happening.

**Comment by matthewr6** [comment]
How often does that happen?  I've noticed it, but it never seemed to happen much and that's never what killed it.  Also, I think the mega food cluster implementation balances that behavior.
I'd like it if someone else looked into that as well, but it may just be an issue of lag or the functions we didn't write, as the nonexistent clusters (which is what those random points are) always disappear in a few seconds.
If needed, though, I can implement the food clusters into master from a different branch, which may or may not fix the problem.

**Comment by an-OK-squirrel** [comment]
I'd like for you guys to do the "using buffers for <100"

**Comment by matthewr6** [comment]
I'll try that tonight if I have time

**Comment by an-OK-squirrel** [comment]
Okay. I still can't get over 50

**Comment by an-OK-squirrel** [comment]
And I can't see what the robot is thinking (which I use a lot for testing)

**Comment by Znapi** [comment]
It loaded for me, but alternating where it is going every frame is a problem. If the direction to the food is the opposite of the random direction, it just ends up sitting there.

**Comment by an-OK-squirrel** [comment]
Exactly.

**Comment by Znapi** [comment]
The problem stops if you remove the else-if in lines [977](https://github.com/ScratchAgarioBots/agario-bot/blob/food-clusters/bot.user.js#L977) to [984](https://github.com/ScratchAgarioBots/agario-bot/blob/food-clusters/bot.user.js#L984). Something must be wrong with the condition you are testing, because it isn't doing what it is supposed to.

**Comment by matthewr6** [comment]
Remove that if/else for now, then.  It's one of the only parts I haven't tested

**Comment by Znapi** [comment]
Should we turn back on enemy buffer rendering?

**Comment by matthewr6** [comment]
What do you mean?

**Comment by Znapi** [comment]
You have buffer rendering commented out in this branch

**Comment by Znapi** [comment]
Here: https://github.com/ScratchAgarioBots/agario-bot/blob/food-clusters/bot.user.js#L762
And some filtering function is commented out there too.

**Comment by an-OK-squirrel** [comment]
Okay

**Comment by Znapi** [comment]
~~Idk, maybe we should leave the filter function commented out. I left the bot running for a while an it's highest score was 105. It was doing better without it for me.~~nvm, this whole time it has been doing good with the filtering on.

**Comment by Znapi** [comment]
Should I push a commit and merge with master? It is working great for me right now(or it is just the server). I got to 230, and to 630 earlier because someone hit a virus and put themselves in a corner.

**Comment by an-OK-squirrel** [comment]
I'm gonna try one more time

**Comment by an-OK-squirrel** [comment]
Bot gets 30 points, other one almost always gets >50...

**Comment by Znapi** [comment]
I got to 230 again, and right now I'm at 182. What is the "other one"?

**Comment by an-OK-squirrel** [comment]
master

**Comment by Znapi** [comment]
Maybe I should try a different server. The one it is in right now is kinda empty.

**Comment by an-OK-squirrel** [comment]
OK

**Comment by an-OK-squirrel** [comment]
Ok, what happened?

**Comment by Znapi** [comment]
It got ~40 twice, then 204. Master is in a slightly more empty server and is at 265.

Aww, it just hit a virus.

**Comment by an-OK-squirrel** [comment]
See? Master is better.

**Comment by Znapi** [comment]
Testing this is so hard. In some servers it is constantly surrounded, and in others it has the whole level to itself.
Master got 65, now it is at 305

**Comment by an-OK-squirrel** [comment]
Servers are always busy for me.

**Comment by Znapi** [comment]
Hit a virus at 484. Should I make a commit anyways, but not merge yet?

**Comment by an-OK-squirrel** [comment]
Sure, i guess. I just feel that it would be best to keep using buffers..

**Comment by Znapi** [comment]
I don't know what is wrong with the super clusters. Maybe it is prioritizing them too much over closer, minor clusters?

I'm going to go for now.

**Comment by an-OK-squirrel** [comment]
BAI

I'll merge after someone does https://github.com/ScratchAgarioBots/agario-bot/pull/49#issuecomment-118065034

**Comment by matthewr6** [comment]
Finished it.  Thinking about taking out mega clusters or moving them to a lower priority, or even having them activate after a certain size (maybe 100?)

**Comment by an-OK-squirrel** [comment]
Still not a good as master right now.

**Comment by matthewr6** [comment]
So what's the difference?  It's just as good as bufferless right now, if not better, and we have to keep moving even if it temporarily gets worse.

**Comment by an-OK-squirrel** [comment]
Maybe it's a problem with the superclusters

**Comment by Znapi** [comment]
How are we defining super clusters? Clusters of clusters?

**Comment by Znapi** [comment]
Part of our problem might be is that when it goes for a cluster/super cluster, it doesn't always pick up all of the food in it. It might pick up one, and then leave, making the long trip to get to the cluster pointless.

It also got to 272, then went off in a random direction and got eaten.

**Comment by an-OK-squirrel** [comment]
Remove superclusters

**Comment by matthewr6** [comment]
Huh, thought I did.  Must not have pushed it up.

**Comment by Znapi** [comment]
I think you did. I am playing with it and I had to uncomment the superclusters.

**Comment by matthewr6** [comment]
Yeah, I did
https://github.com/ScratchAgarioBots/agario-bot/blob/food-clusters/bot.user.js#L950

**Comment by Znapi** [comment]
We should make it so that when it goes towards a food cluster, it picks up all of the food in it by going all the way to the center, and when it goes to a super cluster, it should go to each minor cluster in it one at a time.

**Comment by matthewr6** [comment]
The thing is, once half a minor cluster disappears, the cluster itself is not a cluster anymore.

**Comment by Znapi** [comment]
That is the problem we need to fix. It needs to go all the way to the center, and ignore how the cluster changes. It needs to remember these things inbetween frames, so we should declare some variables outside of `findDestination`. Maybe one variable that stores the location of the next minor cluster it is going to, and then for major clusters, another variable that is an array of minor cluster locations in the next major cluster it is going to.

**Comment by matthewr6** [comment]
Maybe merge this first?  :P  I'll start a new thing tomorrow or Sunday, probably

sorry, I've been busier than I thought I'd be :L

**Comment by an-OK-squirrel** [comment]
Me too :L I don't see why we need to merge though.

**Comment by matthewr6** [comment]
To get it into master...

**Comment by an-OK-squirrel** [comment]
Then make a backup branch :P

**Comment by Znapi** [comment]
What does this have that is different from master? Everything is commented out.

**Comment by an-OK-squirrel** [comment]
Time to compare :P

**Comment by an-OK-squirrel** [comment]
nope

**Comment by Znapi** [comment]
Oh, I forgot about that part. This branch should probably fulfill its own goal before it is merged though.

**Comment by matthewr6** [comment]
It doesn't have a specific goal besides "be better", but it did implement a function to find larger clusters.

**Comment by Znapi** [comment]
I started something for making sure it picks up all of the food in each cluster, and goes to clusters individually, and it is almost done. I just have to fix a small modification I made to the major cluster making function.

**Comment by matthewr6** [comment]
Awesome.  merge it into food-clusters, and then we can merge everything into master.

**Comment by Znapi** [comment]
It turned out I just mispelled a variable name, and it's looking good so far. We should probably test it more thoroughly tomorrow before merging since everyone seems to be busy today, and I haven't tested it too much.

**Comment by matthewr6** [comment]
Cool.

Busy hacking up my Win8.1 computer to have the functionality of a *NIX system but the business workflow of Windows, without having any other OS besides Windows.  Heck yes.

**Comment by Znapi** [comment]
Sounds fun. My dad installed Cygwin on our Windows laptop, but I never could get used to it. I have a Mac anyways, so I just use that.

Should I push a commit or publish another branch now? All that's left is to see if it made improvements. One thing I noticed from running it a few times is that sometimes another blob will eat the cluster the bot is headed to, and so the bot just keeps going for no reason. That and it still runs into enemies. A lot.

**Comment by matthewr6** [comment]
Push a commit if you're checked out to this branch anyways.

Running into enemies doesn't have to do with feeding, I don't think.

Still trying to think of a fix for it.

**Comment by Znapi** [comment]
You're right, it doesn't have to do with feeding. It is just an annoying problem I felt like mentioning.
BTW, I changed the colors that food, clusters, and superclusters are drawn with, so it is easier to tell the difference when I was testing. You can change things back that you don't like if you want.

**Comment by Znapi** [comment]
I'm not sure it helped too much. Its highest score is 144. It doesn't really make a difference until the bot gets big enough that superclusters start appearing. Then if it runs into a crowd of food, it can get big quickly.

**Comment by matthewr6** [comment]
Maybe add something saying if it's less than a certain size, it doesn't go for the mega clusters?  What size is good for that, would you say?

**Comment by Znapi** [comment]
I don't think that is the problem, mainly because the super clusters don't start appearing until it is bigger anyways. It dies a lot when it is small because it is constantly running away from everyone else, and ends up running into them. When it gets bigger, there are less threats, but the enemies are more aggressive because the bot is an ideal size for them to eat.

When the bot was getting the streak to 144, it was in a giant mass of food and without too many enemies around. It did a good job of picking up as much food as it could.

**Comment by matthewr6** [comment]
http://prntscr.com/7otvvo
Let's say you're the green bot.  You're more interested in running away from the left bot (which is faster and more interested in eating you) than from the right bot (going any direction, and it's massive but slow).  The two vectors cancel out the angle, and you just run into the bigger bot.

Seems right from your perspective?

**Comment by Znapi** [comment]
It looks right. I don't see how we can solve these situations(and many others) with vectors.

**Comment by Znapi** [comment]
I'm going to go now. I might think about this later.

**Comment by an-OK-squirrel** [comment]
Right now let's remove megaclusters altogether. I feel that they don't work well enough anyways.

**Comment by matthewr6** [comment]
I'm still thinking only after a certain size.

Vote?

We have removed them completely right now, but we should still keep the function.

**Comment by an-OK-squirrel** [comment]
Remove completely right now.

**Comment by an-OK-squirrel** [comment]
but still keep the function for later :P

**Comment by matthewr6** [comment]
that's what we already did xD

File: bot.user.js [hunk 1]
```diff
@@ -795,41 +853,124 @@ console.log("Running Apos Bot!");
...
+                //var closestFoodCluster;
+                var clusterDist = Infinity;
+
+                /*var*/ foodClusters = getMinorClusters(allClustersAndFood)
+                    .filter(function(cluster) {
+                        var safe = true;
+                        for (var i = 0; i < enemies.length; i++) {
+                            var enemy = enemies[i];
+                            if (enemy.size * 1.1 > player.size * 2) {
+                                if (computeDistance(cluster[0], cluster[1], enemy.x, enemy.y) <= enemy.size + buffer * 2) {
```
Comment (an-OK-squirrel) [inline bot.user.js]: Shouldn't this be bigBufferMultiplier instead of 2?

**Comment by an-OK-squirrel** [comment]
Did I do that comment right?

**Comment by matthewr6** [comment]
Yeah.

File: bot.user.js (hunk 1 above)
Comment (matthewr6) [inline bot.user.js]: Sure

**Comment by an-OK-squirrel** [comment]
Will fix

**Comment by matthewr6** [comment]
Going to go ahead and merge, delete, then make a new `smart-cluster` branch

**Comment by an-OK-squirrel** [comment]
Okay!

**Pull request merged**